src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreBuildHandler.cs

```diff
@@ -0,0 +1,172 @@
+
+            _solutionBuildManager = buildManager;
+        }
+
+        public void Dispose()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            if (_updateSolutionEventsCookieEx != VSCOOKIE_NIL)
+            {
+                ((IVsSolutionBuildManager)_solutionBuildManager).UnadviseUpdateSolutionEvents(_updateSolutionEventsCookieEx);
+                _updateSolutionEventsCookieEx = VSCOOKIE_NIL;
+            }
+        }
+
+        // A factory method invoked internally only
+        internal static async Task<IDisposable> InitializeAsync(Microsoft.VisualStudio.Shell.IAsyncServiceProvider serviceProvider)
+        {
+            Assumes.Present(serviceProvider);
+
+            var instance = new SolutionRestoreBuildHandler();
+
+            var componentModel = await serviceProvider.GetComponentModelAsync();
+            componentModel.DefaultCompositionService.SatisfyImportsOnce(instance);
+
+            await instance.SubscribeAsync(serviceProvider);
+
+            return instance;
+        }
+
+        private async SystemTask SubscribeAsync(Microsoft.VisualStudio.Shell.IAsyncServiceProvider serviceProvider)
+        {
+            // Don't use CPS thread helper because of RPS perf regression
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+            _solutionBuildManager = await serviceProvider.GetServiceAsync<SVsSolutionBuildManager, IVsSolutionBuildManager3>();
+            Assumes.Present(_solutionBuildManager);
+
+            ((IVsSolutionBuildManager6)_solutionBuildManager).AdviseUpdateSolutionEventsEx(this, out _updateSolutionEventsCookieEx);
+        }
+
+        #region IVsUpdateSolutionEvents5
+
+        public void UpdateSolution_QueryDelayBuildAction(uint dwAction, out Guid pActivityId, out IVsTask pDelayTask)
+        {
+            pActivityId = GuidList.guidNuGetSBMEvents;
+            pDelayTask = ThreadHelper.JoinableTaskFactory.RunAsyncAsVsTask(VsTaskRunContext.UIThreadBackgroundPriority, async (token) =>
+            {
+                // move to bg thread to continue with restore
+                await TaskScheduler.Default;
+
+                if ((dwAction & (uint)VSSOLNBUILDUPDATEFLAGS.SBF_OPERATION_CLEAN) != 0 &&
+                    (dwAction & (uint)VSSOLNBUILDUPDATEFLAGS3.SBF_FLAGS_UPTODATE_CHECK) == 0)
+                {
+                    // Clear the project.json restore cache on clean to ensure that the next build restores again
+                    SolutionRestoreWorker.Value.CleanCache();
+                }
+                else if((dwAction & (uint)VSSOLNBUILDUPDATEFLAGS.SBF_OPERATION_BUILD) != 0 &&
+                        (dwAction & (uint)VSSOLNBUILDUPDATEFLAGS3.SBF_FLAGS_UPTODATE_CHECK) == 0 && 
+                        ShouldRestoreOnBuild)
+                {
+                    // check if NuGet lock is already acquired by some other NuGet operation
+                    if (LockService.Value.IsLockHeld)
```

Comment:
yes, but then I don't want to unnecessary acquire lock and release it when I can simply check it.